File: output/GPT3.5-Turbo/xCodeEval_4000/C++/Python/9593a92c2c39eb7502bef10ac7a4dc7c.py
import sys

def getq():
    buf = sys.stdin.readline().strip()
    ret = 0
    for i in range(len(buf)-1, -1, -1):
        ret |= 1 << (ord(buf[i]) - ord('a'))
    return ret

S = sys.stdin.readline().strip()
n = len(S)
S = [ord(c) - ord('a') for c in S]
S.append(31)

m = int(sys.stdin.readline().strip())
Q = [getq() for _ in range(m)]
Qs = set(Q)
ans = {}

stk = []
nn = 0
vis = [False] * 26

for i in range(1, n+1):
    x = S[i]
    if not vis[x]:
        stk.append(x)
        nn += 1
        vis[x] = True
    else:
        j = 0
        while stk[j] != x:
            j += 1
        for k in range(j, nn-1):
            stk[k] = stk[k+1]
        stk[nn-1] = x
    t = 0
    for j in range(nn-1, -1, -1):
        t |= 1 << stk[j]
        if t & (1 << S[i+1]):
            break
        if t in Qs:
            ans[t] = ans.get(t, 0) + 1

for i in range(m):
    print(ans.get(Q[i], 0))

File: output/GPT3.5-Turbo/xCodeEval_4000/C++/Java/9593a92c2c39eb7502bef10ac7a4dc7c.java
import java.util.*;
import java.io.*;

class Main {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        String S = br.readLine();
        int n = S.length();
        int[] arr = new int[n + 1];
        for (int i = 1; i <= n; i++) {
            arr[i] = S.charAt(i - 1) - 'a';
        }
        arr[n + 1] = 31;
        int m = Integer.parseInt(br.readLine());
        int[] Q = new int[m + 1];
        Set<Integer> Qs = new HashSet<>();
        Map<Integer, Integer> ans = new HashMap<>();
        for (int i = 1; i <= m; i++) {
            Qs.add(Q[i] = getq(br));
        }
        int[] stk = new int[27];
        int nn = 0;
        boolean[] vis = new boolean[26];
        for (int i = 1; i <= n; i++) {
            int x = arr[i];
            if (!vis[x]) {
                stk[++nn] = x;
                vis[x] = true;
            } else {
                int j;
                for (j = 1; ; j++) {
                    if (stk[j] == x) break;
                }
                for (; j < nn; j++) {
                    stk[j] = stk[j + 1];
                }
                stk[nn] = x;
            }
            int t = 0;
            for (int j = nn; j > 0; j--) {
                t |= 1 << stk[j];
                if ((t & (1 << arr[i + 1])) != 0) {
                    break;
                }
                if (Qs.contains(t)) {
                    ans.put(t, ans.getOrDefault(t, 0) + 1);
                }
            }
        }
        for (int i = 1; i <= m; i++) {
            System.out.println(ans.get(Q[i]));
        }
    }

    public static int getq(BufferedReader br) throws IOException {
        String buf = br.readLine();
        int ret = 0;
        for (int i = buf.length() - 1; i >= 0; i--) {
            ret |= 1 << (buf.charAt(i) - 'a');
        }
        return ret;
    }
}
